File: 41.缺失的第一个正数.cc
#include <vector>
#include <algorithm>
#include <cmath>
using std::vector;
using std::abs;

class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        int n = nums.size();
        for (int& x : nums)
            if (x <= 0)
                x = n + 1;
        for (int i = 0; i < n; ++i) {
            int x = abs(nums[i]);
            if (x <= n)
                nums[x - 1] = -abs(nums[x - 1]);
        }
        for (int i = 0; i < n; ++i)
            if (nums[i] > 0)
                return i + 1;
        return n + 1;
    }
};

File: TypeScript/41.缺失的第一个正数.ts
function firstMissingPositive(nums: number[]): number {
    const n = nums.length
    nums = nums.map(x => x <= 0 ? n + 1 : x)
    for (let i = 0; i < n; ++i) {
        const x = Math.abs(nums[i])
        if (x <= n)
            nums[x-1] = -Math.abs(nums[x-1])
    }
    for (let i = 0; i < n; ++i) {
        if (nums[i] > 0)
            return i + 1
    }
    return n + 1
};
